### ISSUE TITLE
Infinite recursion is not catched by the compiler.

### ISSUE THREAD

**Issue by wdanilo** [rust-lang/rust#70727]
Hi, consider this code:

```rust
#![deny(unconditional_recursion)]

struct Object {
    
}

impl Object {
    // RENAME HERE:
    fn foo(&self) -> f32 {
        7.0
    }
}

impl ToObject for Object {
    fn to_object(&self) -> &Object {
        self
    }
}

trait ToObject {
    fn to_object(&self) -> &Object;
}


trait ObjectOps : ToObject {
    fn foo(&self) -> f32 {
        self.to_object().foo()
    }
}

impl<T:ToObject> ObjectOps for T {}

fn main() {
    let x = Object{};
    println!("{}",x.foo());
}
```

It works fine. Let's rename `foo` to `foo2` next to the comment "RENAME HERE". What happens is that this code has infinite recursion now and `rustc` doesn't report any error nor warning despite the usage of `#![deny(unconditional_recursion)]`. 

This is a serious problem, as a simple refactoring can cause something like that and it is very, very hard to debug, especially on WASM platform, where infinite recursion can just return a random number (it is UB). This problem could be solvable if I was able to write somehow `MAGIC::Object::foo(self.to_object())` instead of `self.to_object().foo()` but in such a way, that the methods from `ObjectOps` would not be visible in `MAGIC::Object`. Currently, when refactoring such code we can get infinite recursion and spend hours debugging it :(

**Comment by jumbatm** [comment]
Thanks for the report!

Very slightly trimmed test case:

```rust
#![deny(unconditional_recursion)]

struct Object;

trait ToObject {
    fn to_object(&self) -> &Object;
}

impl ToObject for Object {
    fn to_object(&self) -> &Object {
        self
    }
}

impl<T: ToObject> ObjectOps for T {}

trait ObjectOps : ToObject {
    fn foo(&self) -> f32 {
        self.to_object().foo()
    }
}

fn main() {
    let x = Object;
    println!("{}", x.foo());
}
```
The (false?) negative for `unconditional_recursion` essentially boils down to this case:

```rust
#![deny(unconditional_recursion)]
fn bar() {
    foo();
}

fn foo() {
    bar();
}

fn main() {
    foo();
}
```
ie, when the cycle in the callgraph is formed with at least one other function. This is because the lint currently only fires if a function directly calls itself, eg:

```rust
fn foo() {
    foo();
}
```
Relevant comment here:

https://github.com/rust-lang/rust/blob/548afdbe1a600ba868ac00acc4e94cca0242b001/src/librustc_mir_build/lints.rs#L29-L37

**Comment by jonas-schievink** [comment]
Part of https://github.com/rust-lang/rust/issues/57965

**Comment by etherealist** [comment]
Is this the same issue or should I open an extra issue:

```rust
struct Floaty {}

impl From<f64> for Floaty {
    fn from(value: f64) -> Floaty {
        value.into()
    }
}

fn main() {
    let _: Floaty = 1.0.into();
}
```

This does not generate any recursion warnings.

On musl any recursions segfault, which I reported here: #75667. The above was my original case, util I realized that it segfaults for any recursions.

**Comment by jonas-schievink** [comment]
@etherealist Yeah, same issue

**Comment by jhg** [comment]
Maybe this is also same issue:
```
use std::fmt;

#[derive(Copy)]
pub enum MyEnum {
    A,
    B,
}

impl AsRef<str> for MyEnum {
    fn as_ref(&self) -> &str {
        match self {
            Self::A => "a",
            Self::B => "b",
        }
    }
}

impl fmt::Display for MyEnum {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", &self)
    }
}
```
Maybe it is calling recursively to Display::fmt but the compiler does not detect it is calling to itself when it is a trait.

**Comment by Ren-ZY** [comment]
Hello, guys!
I have done some work on this problem.
I developed a tool named "RUSD" in my repo: https://github.com/Ren-ZY/RUSD
This tool first build a call-graph (support simple generic) of a whole crate on MIR and then use Tarjan+Johnson Algorithm to locate all the cycles in the call-graph, no matter these cycles have intersections or not.
In other words, it can deal with cross-function recursion.
I have test this tool in six Rust stackoverflow CVEs and can find all the problematic recursive functions.
Welcome to use our tool to locate those troublesome recursive functions！
